1.train_models/log_reg_plates_3_3p_5_cp_norm_data/nbconverted/plate_3_3p_5_train_log_reg_scv.py

```diff
@@ -0,0 +1,365 @@
+
+
+# ## Encode genotypes and extract feature data
+
+# In[8]:
+
+
+le = LabelEncoder()
+
+y = le.fit_transform(restdf["Metadata_genotype"])
+X = restdf.drop(columns=meta_cols)
+
+
+# # Train Models
+
+# ## Specify parameters for training
+
+# In[9]:
+
+
+logreg_params = {
+    "max_iter": 250,
+    "random_state": 0,
+    "n_jobs": -1,
+    "penalty": "l2",
+}
+
+# Random sampling range of hyperparameter
+param_ranges = {
+    "C": (0, 200)
+}
+
+# Number of iteration to optimize hyperparameters
+rand_iter = 500
+
+# Best accuracy
+best_acc = 0
+
+# Initial accuracy
+acc = 0
+
+# Number of folds
+n_splits = 8
+
+
+# Generate hyperparameter samples
+random_params = {
+    i:
+    {key: random.uniform(*param_ranges[key]) for key in param_ranges}
+    for i in range(rand_iter)
+}
+
+
+# ## Hyperparameter search
+
+# In[10]:
+
+
+# Iterate through hyperparameters
+for idx, rparams in random_params.items():
+
+    skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=0)
+
+    # Combine parameters in current search with logistic regresssion parameters
+    comb_params = logreg_params | rparams
+
+    hparam_data = 0
+    hparam_data = defaultdict(dict)
+
+    # Loop through the folds
+    for fold, (train_index, val_index) in enumerate(skf.split(X, y)):
+
+        # Store split data for folds
+        hparam_data[f"fold{fold}"]["train_idx"] = train_index
+        hparam_data[f"fold{fold}"]["val_idx"] = val_index
+
+        X_train, X_val = X.iloc[train_index], X.iloc[val_index]
+        y_train, y_val = y[train_index], y[val_index]
+
+        # Prevent the convergence warning in sklearn
+        with parallel_backend("multiprocessing"):
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    "ignore", category=ConvergenceWarning, module="sklearn"
+                )
+                logreg = LogisticRegression(**comb_params)
+                logreg.fit(X_train, y_train)
+
+        # Cumulative accuracy for all folds
+        preds = logreg.predict(X_val)
+        acc += accuracy_score(y_val, preds)
+
+    # Average accuracy for the folds
+    acc = acc / n_splits
+
+    # Store the data with the best performance
+    if acc > best_acc:
+        best_hparam = hparam_data.copy()
+        best_acc = acc
+        best_hp = rparams
+
+print(f"Best average validation accuracy = {best_acc}")
+
+
+# ## Retrain model
+
+# In[11]:
+
+
+logreg_params = {
+    "max_iter": 3000,
+    "random_state": 0,
+    "n_jobs": -1,
+    "penalty": "l2",
+}
+
+comb_params = logreg_params | best_hp
+
+logreg = LogisticRegression(**comb_params)
+logreg.fit(X, y)
+
+
+# # Save models and model data
+
+# ## Save model
+
+# In[12]:
+
+
+data_suf = "log_reg_cp_fs_data_plate_5"
+
+# Save the models
+dump(logreg, f"{models_path}/{data_suf}.joblib")
+
+# Save label encoder
+dump(le, f"{data_path}/label_encoder_{data_suf}.joblib")
```

Comment:
Double checking, should this be stored in the model or perhaps a "methods" dir to help distinguish between data? Feel free to dismiss if this doesn't add value.

Reply:
It could, but I think the model would also fall under this category of "objects saved containing subroutines". I don't think there is a rigid distinction here